### PR TITLE
[TestFramework] Added prompt to install the test-proxy tool

### PR DESCRIPTION
**Pull request by kinelski**
This PR updates the Test Framework to automatically install the `test-proxy` tool when running in Record mode.

### CODE REVIEW

File: sdk/core/Azure.Core.TestFramework/src/RecordedTestBase.cs
Comment (pallavit) [inline sdk/core/Azure.Core.TestFramework/src/RecordedTestBase.cs:366]: Do we also need to update the documentation for this new scenario? If so, what would be the right place to do it?

/cc: @benbp , @weshaggard do we have central documentation around this for partners? Or each language is making independent calls here?

File: eng/scripts/Push-TestProxy.ps1
Comment (JoshLove-msft) [inline eng/scripts/Push-TestProxy.ps1:21]: Can we add instructions for how they would do it themselves if they respond with No?

File: sdk/core/Azure.Core.TestFramework/src/RecordedTestBase.cs
Comment (christothes) [inline sdk/core/Azure.Core.TestFramework/src/RecordedTestBase.cs:389]: rather than having the TestFramework delete these, would it be better to log a warning if we detect that the assets are migrated to the proxy test recording repo and tell them it is safe to delete them manually?

Comment (JoshLove-msft) [reply sdk/core/Azure.Core.TestFramework/src/RecordedTestBase.cs:389]: This is existing code for deleting unused recordings. I think it was just refactored to remove one level of nesting, so the diff is misleading.

Comment (kinelski) [reply sdk/core/Azure.Core.TestFramework/src/RecordedTestBase.cs:389]: > This is existing code for deleting unused recordings. I think it was just refactored to remove one level of nesting, so the diff is misleading.

That's correct. Not part of this PR.

File: eng/scripts/Push-TestProxy.ps1
Comment (scbedd) [reply eng/scripts/Push-TestProxy.ps1:1]: I'm not opposed at all to making the update operation that we call in `eng/common/testproxy/test-proxy-tool.yml` this script. I'll file an issue and dash out a PR for that this week. Not a big change at all.

However, aside from the installation part, I don't see much benefit in parameterizing `test-proxy push` as the arguments are pretty much exactly the same as what you pass to this script here. (the assets.json).

File: eng/scripts/Push-TestProxy.ps1
Comment (JoshLove-msft) [reply eng/scripts/Push-TestProxy.ps1:21]: I don't think they would need to install the tool, so we can skip that part.

File: sdk/core/Azure.Core.TestFramework/src/RecordedTestBase.cs
Comment (scbedd) [reply sdk/core/Azure.Core.TestFramework/src/RecordedTestBase.cs:366]: @pallavit Are you referring to `asset-sync` in general? EG externalizing the recordings? if the partners are contributing code to this (or other azure-sdk-for-X) repos, I think they should probably follow what the rest of the folks in this repo are doing and store their data in `azure-sdk-assets`.

Anyone part of the `azure-sdk` umbrella (this includes the azure-sdk-write-X or azure-sdk-push-X teams) has write access to the asets repo.

Other than that:
- java auto downloads the standalone executable and leaves it up to the users to push
- js integrated the migration script + push/restore commands into their utility recorder package (version of azure.core.test)
- python added the proxy push/restore to their local dev script
- C++ run it manually as far as I understand it
- Go does the same

File: sdk/core/Azure.Core.TestFramework/src/TestRecordingMismatchException.cs
Comment (kinelski) [inline sdk/core/Azure.Core.TestFramework/src/TestRecordingMismatchException.cs:30]: It may be a good idea to add a link to the Test Framework docs once they're merged.

File: sdk/core/Azure.Core.TestFramework/README.md
Comment (pallavit) [inline sdk/core/Azure.Core.TestFramework/README.md:223]: Do we by-default ignore .assets folder from showing in git commands? If not, should we do that to prevent accidental commit of these folders via git add * etc.?

Comment (pallavit) [reply sdk/core/Azure.Core.TestFramework/README.md:223]: I see .gitignore having .assets in its list, so we are good here!

File: sdk/core/Azure.Core.TestFramework/README.md
Comment (JoshLove-msft) [inline sdk/core/Azure.Core.TestFramework/README.md:235]: Do we really need this setting? We should be able to detect whether the tool is installed.

File: sdk/core/Azure.Core.TestFramework/src/RecordedTestBase.cs
Comment (JoshLove-msft) [inline sdk/core/Azure.Core.TestFramework/src/RecordedTestBase.cs:577]: I'm not sure that we need the local support. All of our docs recommend installing globally. If someone has installed it locally, they should still just install it globally as well.

File: sdk/core/Azure.Core.TestFramework/src/TestRecordingMismatchException.cs
Comment (JoshLove-msft) [inline sdk/core/Azure.Core.TestFramework/src/TestRecordingMismatchException.cs:30]:
```suggestion
            const string Reminder = "If this is a new recording, make sure you have pushed the recordings to the assets repository.";
```

File: sdk/core/Azure.Core.TestFramework/README.md
Comment (kinelski) [reply sdk/core/Azure.Core.TestFramework/README.md:235]: I think I'm just being overconcerned at this point. I'll keep it simple for now and add the setting back if we ever need it.

File: sdk/core/Azure.Core.TestFramework/src/RecordedTestBase.cs
Comment (JoshLove-msft) [inline sdk/core/Azure.Core.TestFramework/src/RecordedTestBase.cs:455]: Hmm so this will prompt everyone immediately once this is merged but before their assets are migrated. Should we make this dependent on whether the assets are migrated?

Comment (kinelski) [reply sdk/core/Azure.Core.TestFramework/src/RecordedTestBase.cs:455]: There's a check in `AssertTestProxyToolIsInstalled` that skips the prompt if `AssetsJsonPath` is equal to null. I believe this does what you're suggesting.

File: sdk/core/Azure.Core.TestFramework/README.md
Comment (JoshLove-msft) [inline sdk/core/Azure.Core.TestFramework/README.md:223]: Do we need to condition these docs on whether the migration has taken place yet?

File: sdk/core/Azure.Core.TestFramework/README.md
Comment (JoshLove-msft) [inline sdk/core/Azure.Core.TestFramework/README.md:227]: Same comment here on whether we need to say this is only for migrated services.

File: sdk/core/Azure.Core.TestFramework/README.md
Comment (kinelski) [reply sdk/core/Azure.Core.TestFramework/README.md:223]: Since we intended to start moving recordings right after merging this PR, my original intent was to completely replace the docs with new ones. After today's discussion we're still not sure how things will be on the management side, so I'll take your suggestion and explain the two scenarios.

Comment (kinelski) [reply sdk/core/Azure.Core.TestFramework/README.md]: Updated.

File: sdk/core/Azure.Core.TestFramework/src/TestRecordingMismatchException.cs
Comment (JoshLove-msft) [inline sdk/core/Azure.Core.TestFramework/src/TestRecordingMismatchException.cs:33]: nit: why 2 new lines?

**Review by JoshLove-msft — APPROVED** [review]
Looks great!

**Review by annelo-msft — APPROVED** [review]
I'm interested to learn more about what this UX will be like for us in our daily flow - thanks for some really nice written explanations to help us and our partners understand!

File: sdk/core/Azure.Core.TestFramework/src/TestRecordingMismatchException.cs
Comment (JoshLove-msft) [reply sdk/core/Azure.Core.TestFramework/src/TestRecordingMismatchException.cs:33]: Fair enough!

**Pull request merged**
